ledger-service/http-json/src/main/scala/com/digitalasset/http/domain.scala

```diff
@@ -252,32 +262,159 @@ package domain {
+      }
+    }
+    object Arguments {
+      final case class Blob(payloadBlob: PbAny) extends Arguments[Nothing]
+      final case class Record[+LfV](payload: LfV) extends Arguments[LfV]
+      private[http] val blobKey = "payloadBlob"
+      private[http] val recordKey = "payload"
+
+      implicit val covariant: Traverse[Arguments] = new Traverse[Arguments] {
+        override def traverseImpl[G[_]: Applicative, A, B](
+            fa: Arguments[A]
+        )(f: A => G[B]): G[Arguments[B]] =
+          fa match {
+            case b @ Blob(_) => Applicative[G] point b
+            case Record(payload) => f(payload) map (Record(_))
+          }
+      }
+    }
+
+    final case class Metadata(
+        createdAt: Time.Timestamp,
+        contractKeyHash: Option[Base64],
+        driverMetadata: Option[Base64],
+    ) {
+      import com.daml.api.util.TimestampConversion.{fromLf => timestampLfToProto}
+
+      def toLedgerApi: lav1.contract_metadata.ContractMetadata =
+        lav1.contract_metadata.ContractMetadata(
+          createdAt = Some(timestampLfToProto(createdAt)),
+          contractKeyHash = contractKeyHash.cata(Base64.unwrap, ByteString.EMPTY),
+          driverMetadata = driverMetadata.cata(Base64.unwrap, ByteString.EMPTY),
+        )
+    }
+
+    object Metadata extends ((Time.Timestamp, Option[Base64], Option[Base64]) => Metadata) {
+      // meant only for testing
+      private[http] def fromLedgerApi(
+          la: lav1.contract_metadata.ContractMetadata
+      ): Error \/ Metadata = {
+        import com.daml.api.util.{TimestampConversion => TSC}
+        def boxIfNE(bs: ByteString) = if (bs.isEmpty) None else Some(Base64(bs))
+        for {
+          createdAt <- (la.createdAt toRightDisjunction
+            Error(Symbol("Metadata.fromLedgerApi"), "missing createdAt field"))
+        } yield Metadata(
+          TSC.toLf(createdAt, TSC.ConversionMode.Exact),
+          boxIfNE(la.contractKeyHash),
+          boxIfNE(la.driverMetadata),
+        )
+      }
+    }
+
+    implicit val covariant: Bitraverse[DisclosedContract] =
+      new Bitraverse[DisclosedContract] {
+        override def bitraverseImpl[G[_]: Applicative, A, B, C, D](
+            fab: DisclosedContract[A, B]
+        )(f: A => G[C], g: B => G[D]): G[DisclosedContract[C, D]] =
+          ^(f(fab.templateId), fab.arguments traverse g) { (templateId, arguments) =>
+            fab.copy(templateId = templateId, arguments = arguments)
+          }
+      }
+  }
+
+  /** @tparam TmplId disclosed contracts' template ID
+    * @tparam LfV disclosed contracts' payload encoding
+    */
+  final case class CommandMeta[+TmplId, +LfV](
       commandId: Option[CommandId],
       actAs: Option[NonEmptyList[Party]],
       readAs: Option[List[Party]],
       submissionId: Option[SubmissionId],
       deduplicationPeriod: Option[domain.DeduplicationPeriod],
+      disclosedContracts: Option[List[DisclosedContract[TmplId, LfV]]],
```

Comment:
Would it make things simpler if this were
```suggestion
      disclosedContracts: Option[List[lav1.commands.DisclosedContract]],
```
i.e. the protobuf value? 

AFAICT we don't actually inspect the contents of the DisclosedContracts except for conversions from json->domain->proto and this would mean we wouldn't need to care about and thread through the template types.

In the JSON I could imagine these being json-encoded protobuf, which we should be able to decode quite simply.

Reply:
As @cocreature suggests, this would introduce a separate, incompatible encoding of contract payloads, that would be quite complex to assemble without client-side library support. Currently, while we encourage JavaScript users to at least use `@daml/ledger`, it is quite possible to get by with only an HTTP and JSON library.

Moreover, there is a use-case consideration: since disclosed contracts is meant to support out-of-band disclosure, I think it makes most sense to significantly weigh use cases where encoded form _does not_ precisely flow isomorphism-to-isomorphism from a data response to the discloser, to a data request from the disclosee. The upshot being that clients must assemble disclosed contracts themselves, and then we come back to the point about client-side library support.